### PR TITLE
[FLINK-25256][streaming] Externally induced sources replay barriers received over RPC instead of inventing them out of thin air. 

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

Externally induced sources are currently ill-defined and have a lot of unnecessary limitations. This PR addresses it by explicitly only holding back checkpoint barriers until the external source induces it. 

While this approach seemingly limits the way externally induce sources work (they can't trigger a checkpoint on their own anymore), it actually explicitly supports the only plausible way. Sources simply can't trigger a checkpoint on their own - the checkpoint coordinator needs to track it and multiple sources need the coordinator to work at all (or else they deadlock).

## Brief change log

- Externally induced sources replay barriers received over RPC instead of inventing them out of thin air.
- Clarify the contract.
- Migrate related tests to JUnit5 and AssertJ.

## Verifying this change

Expanded the original unit test to check that the barrier is correctly relayed.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java:131]: Not sure about the performance implications here. In general, why don't we process all RPC directly in the mailbox?

Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java:131]: We kind of do. The only check so far outside of the mailbox was a local flag check `isSynchronous(checkpointOptions.getCheckpointType())`. All the rest is immediately executed in the mailbox.

As for the change, the way I see it now, is that triggering a checkpoint will require two mailbox actions instead of one.
One for checking the flags (`isSynchronous`/`externallyInduced` or so) and the second one for executing the actual trigger. 

I am a bit hesitant about such a change that affects all jobs for the sake of really rare ones that use `externallyInducedSources`.

The alternatives I see are either 1) rework `triggerStopWithSavepointAsync`, `triggerCheckpointAsync` to not submit a second action, which IMO would not be trivial and poses some risks or 2) execute the logic for externally induced sources in the mailbox and act on its results. Sth like:

```
enum TriggerAction {
IMMEDIATE,
DELAYED
}

public CompletableFuture<Boolean> triggerCheckpointAsync(
          CheckpointMetaData checkpointMetaData, CheckpointOptions checkpointOptions) {
    if (isExternallyInducedSource) {
      return triggerExternallyInducedSourcesCheckpointAsync()
        .thenCompose(action -> {
              switch (action) {
                     case IMMEDIATE:
                         return doTriggerCheckpointAsync();
                     case DELAYED:
                         return CompletableFuture.completedFuture(isRunning());
              }
          })
    } else {
      return doTriggerCheckpointAsync(...);
    }
}

CompletableFuture<TriggerAction> triggerExternallyInducedSourcesCheckpointAsync() {
mainMailboxExecutor.execute(...)
return action;
}

CompletableFuture<Boolean> doTriggerCheckpointAsync(CheckpointMetaData checkpointMetaData, CheckpointOptions checkpointOptions) {
...
}

```

Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java:131]: I quickly reworked it, such that it only needs 2 mails for the rare case of externally induced sources. WDYT?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java:206]: We should probably discuss if this is the best choice.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java:134]: The cleanup (and the one in #trigger) don't work well with concurrent checkpoints. Do we have a way to determine max concurrent checkpoints or can we actually rely on `abortCheckpoint`?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java
Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java:131]: But it still requires 2 mails for checkpoints without externally induced sources, doesn't it?

Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java:134]: my bad, I didn't force push -.-

Comment (dawidwys) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java:134]: That looks actually very similar to the code I wrote ;) so I think I am good with it.

I will continue checking the PR as I have not read through everything.

Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java:134]: Yes, it's heavily inspired by your idea but I didn't get the idea behind `TriggerAction`.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java:130]: Doesn't it cause hot looping? It also adds an additional condition on the hot path, doesn't it?

Would it be possible to `suspendDefaultAction` when we add something to the `triggeredCheckpoints` and unfreeze it once the `triggeredCheckpoints` are empty?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java:255]: Shouldn't we do it in the mailbox? We are modifying fields that are accessed from RPC and the mailbox.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java:130]: Yes, I wanted to avoid it for complexity but I guess, we need to do it. That probably means tinkering with the avail future.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java:255]: Good catch. I keep forgetting that RPC are not in mailbox. :/
I added assertions now and called them over mailbox in the async RPC calls. (`declineCheckpoint` is already in mailbox).

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java]: I beefed `StreamTaskExternallyInducedSourceInput` a bit to not impact the hot-path of normal sources. PTAL.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java:271]: Shouldn't we potentially unblock the input here? If the only pending checkpoint was `aborted/declined/cancelled`?

**Review by dawidwys — COMMENTED** [review]
How hard would it be to add a test for the blocking/unblocking of the externally induced source?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java:271]: You are absolutely right.

**Comment by AHeise** [comment]
> How hard would it be to add a test for the blocking/unblocking of the externally induced source?

I have added assertions into the main test method that cover that. Please check if you think I should have additional test cases.

**Comment by AHeise** [comment]
This PR has been verified by the flink-pravega maintainers to work on their tests for checkpoints (savepoint test pending).

**Review by crazyzhou — COMMENTED** [review]
I have tested with basic Pravega reader application savepointing locally with a rocksdb state backend, with a simple app
job graph:
image

After this fix, this app can successfully do the `stop-with-savepoint` while it failed before.

It can recover nicely with the savepoint: 
image

The application is now only having `_metadata` file in each checkpoint and savepoint, so still trying some more complicated cases to see I can reproduce the issue.

**Comment by dawidwys** [comment]
On the matter of writing into separate files instead of keeping data inside of the metadata, you might want to have a look at: `state.storage.fs.memory-threshold`

**Review by crazyzhou — APPROVED** [review]
I have tested with an app with a larger state, here is the savepoint structure:
```
:~/flink-savepoints/savepoint-9a7e45-0215c1d6f545$ ls
9c355659-2475-4402-a5e7-3450c70394b4  _metadata
```

The application with Pravega source can both cancel and stop with savepoint nicely, and can successfully recover from that.

**Pull request merged**
